Source/WebCore/style/StyleBuilderConverter.h

```diff
@@ -879,32 +878,41 @@ inline RefPtr<QuotesData> BuilderConverter::convertQuotes(BuilderState&, const C
     return QuotesData::create(quotes);
 }
 
-inline TextUnderlinePosition BuilderConverter::convertTextUnderlinePosition(BuilderState&, const CSSValue& value)
+inline static OptionSet<TextUnderlinePosition> valueToUnderlinePosition(const CSSPrimitiveValue& primitiveValue)
 {
-    auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value);
-    if (primitiveValue) {
-        switch (primitiveValue->valueID()) {
-        case CSSValueFromFont:
-        case CSSValueUnder:
-            return { fromCSSValueID<TextUnderlinePosition::Metric>(primitiveValue->valueID()), TextUnderlinePosition::Side::Auto };
-
-        case CSSValueLeft:
-        case CSSValueRight:
-            return { TextUnderlinePosition::Metric::Auto, fromCSSValueID<TextUnderlinePosition::Side>(primitiveValue->valueID()) };
+    ASSERT(primitiveValue.isValueID());
 
-        default:
-            return { TextUnderlinePosition::Metric::Auto, TextUnderlinePosition::Side::Auto };
-        }
+    switch (primitiveValue.valueID()) {
+    case CSSValueFromFont:
+        return TextUnderlinePosition::FromFont;
+    case CSSValueUnder:
+        return TextUnderlinePosition::Under;
+    case CSSValueLeft:
+        return TextUnderlinePosition::Left;
+    case CSSValueRight:
+        return TextUnderlinePosition::Right;
+    case CSSValueAuto:
+        return RenderStyle::initialTextUnderlinePosition();
+    default:
+        break;
     }
 
+    ASSERT_NOT_REACHED();
+    return RenderStyle::initialTextUnderlinePosition();
+}
+
+inline OptionSet<TextUnderlinePosition> BuilderConverter::convertTextUnderlinePosition(BuilderState&, const CSSValue& value)
+{
+    if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value))
+        return valueToUnderlinePosition(*primitiveValue);
+
     auto* pair = dynamicDowncast<CSSValuePair>(value);
     if (!pair)
-        return { TextUnderlinePosition::Metric::Auto, TextUnderlinePosition::Side::Auto };
+        return { };
 
-    return {
-        fromCSSValueID<TextUnderlinePosition::Metric>(downcast<CSSPrimitiveValue>(pair->first()).valueID()),
-        fromCSSValueID<TextUnderlinePosition::Side>(downcast<CSSPrimitiveValue>(pair->second()).valueID()),
-    };
```

Comment:
Not sure it's better, but we can construct it directly.

`return { valueToUnderlinePosition(...pair->first) , valueToUnde...(pair->second) }`

Reply:
That doesn't work because `valueToUnderlinePosition` returns an `OptionSet`, so what you're suggesting would be initializing nested OptionSets

Reply:
Oh yes we need ` return valueToUnder(first) & valueToUnder(second);`, which is definitely not better 😁